### PR TITLE
Enhance accessibility of autocomplete

### LINKED ISSUE #4617: Improve accessibility of the autocomplete plugin (divarea editor only)
## Type of report

Bug

## Provide detailed reproduction steps (if any)

Extracted from #2199:

> Autocomplete accessibility markup should be reviewed. Items should gain proper roles etc.

Since there is no way to provide support in a reasonable way for `iframe` based editor (classic editor), see https://github.com/ckeditor/ckeditor4/issues/2199#issuecomment-818711934, we can cover this for `divaera` / `inline` one at least.

### LINKED ISSUE #4653: Autocomplete's change-isActive event is incorrectly propagated
## Type of report

Bug

## Provide detailed reproduction steps (if any)

1. Open console
2. Go to https://jsfiddle.net/Comandeer/qy2gmcLh/
3. Open autocomplete by typing `@`.
4. Close autocomplete by pressing <kbd>Esc</kbd>

### Expected result

There is the following sequence logged:

```
1 true
2 true
1 false
2 false
```

### Actual result

The following sequence is logged:

```
1 true
2 true
```

## Other details

* CKEditor version: 4.10+
* Installed CKEditor plugins: autocomplete

It seems to be caused by #4652. There are two possible solutions:

* fix #4652 – which shouldn't be _that_ risky change,
* change the event data passed to the `change-isActive` to an object, which is a breaking change and can break existing integrations (e.g. [in SCAYT](https://github.com/WebSpellChecker/ckeditor-plugin-scayt/blob/master/plugin.js#L30-L43))

### PR DESCRIPTION
**Pull request by Comandeer**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#4617](https://github.com/ckeditor/ckeditor4/issues/4617): Fixed: [Autocomplete](https://ckeditor.com/cke4/addon/autocomplete) is not accessible in inline editors.
```

## What changes did you make?

I've added appropriate attributes to the editor's editable (`[aria-autocomplete="list"]`, `[aria-controls]`, `[aria-activedescendant]`, `[aria-expanded]`) and to the autocomplete and its items (mainly roles – `listbox` for the autocomplete itself and `option` for its items). Thanks to that screen readers far better support the autocomplete:

* NVDA announces the presence of the autocomplete on editor focus and reads options from the listbox,
* JAWS announces the presence of the autocomplete on editor focus if the editor is empty (it's probably connected with the default verbosity settings) and reads options from the listbox,
* VoiceOver does _not_ announce the presence of the autocomplete, however it reads options from the listbox.

I've also ensured that these changes aren't applied to iframe-based editors.

## Which issues does your PR resolve?

Closes #4617.
Closes #4653.

### CODE REVIEW

**Comment by sculpt0r** [comment]
Rebase on newest `master`.

**Review by sculpt0r — CHANGES_REQUESTED** [review]
@Comandeer 

I added a few inline comments, but not finish the entire review since we have red CI for `a11y` tests (https://travis-ci.com/github/ckeditor/ckeditor4/jobs/501340888). I re-run it once for both Chrome builds with the same result.

Only the second Chrome build is red. I run `a11y` manually on my Chrome & FF (also on BS Chrome & FF) and it looks good :thinking: Also that is a case with unexpected errors in CI logs (https://github.com/ckeditor/ckeditor4/issues/4620).

Could you take a look at `a11y` tests?

File: plugins/autocomplete/plugin.js [hunk 1]
```diff
@@ -296,13 +299,15 @@
 			// Register keybindings if editor is already initialized.
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/autocomplete/plugin.js:299]: Instead of moving it, I'll rephrase it to better match the current code.

**Comment by Comandeer** [comment]
I've updated methods' names.

As for tests: they work for me :thinking: I don't know why they are failing on CI.

**Review by sculpt0r — CHANGES_REQUESTED** [review]
Thank you for the previous changes.

I made some research in a11y unit tests and here are my results:
- first of all, those tests are not independent. If one fails, then another may also fail. I found it when I try to add assertion with spy on the method which updates aria attributes on editable. I put it on a separate branch based on yours: https://github.com/ckeditor/ckeditor4/pull/4647/commits/987869614cfdbfcc11c1c5a1584fa6e821f2ca5c The spy assertion fails on my PC, but it causes another test to fail... You can verify it even by adding a simple `assert.isTrue( false );`. Finally, I realized that if an assertion fails, then `autoComplete.destroy();` is not invoked... So I made a fast modification here: https://github.com/ckeditor/ckeditor4/commit/82d5147b1a721b3e1d9ffb114592b97ca2eb88a8 and it leads to only two failing tests on CI (was four before): https://travis-ci.com/github/ckeditor/ckeditor4/jobs/501698276

Could you make a little deeper digging with those two? Also please make sure, that created autoComplete will be destroyed after each test. We won't have a situation when one test influences another.

Back to https://github.com/ckeditor/ckeditor4/pull/4647/commits/987869614cfdbfcc11c1c5a1584fa6e821f2ca5c - I supposed that `sinon.spy( autoComplete, 'updateAriaAttributesOnEditable' ),` and assertion `assert.isTrue( ariaUpdateSpy.calledWith( false ) );` whould pass after `ESC` key press simulation, but it doesn't...

So we have only two failing tests on CI. The code behavior seems to be valid (especially it pass on local machines). So it must be something 'deeper'. Based on the above, maybe it's test construction? It's definitely a problem only with those new tests - so again, please dig a little big deeper.

**Comment by Comandeer** [comment]
It seems that tests fail due to #4653 – `change-isActive` is incorrectly handled as `evt.data` is set to `false`.

**Comment by f1ames** [comment]
I guess it's now unblocked since #4652 was fixed?

**Comment by Comandeer** [comment]
Rebased onto latest `master`.

**Comment by Comandeer** [comment]
Ok, it seems that fix for #4652 fixed issues with tests. I've also added a small test for #4653.

**Review by sculpt0r — CHANGES_REQUESTED** [review]
@Comandeer 
It's nice that you solved an issue with failing CI - which actually was a manifestation of another issue with our events :+1: Now we have an even better code base after all these years :)

Screen readers announce autocomplete in a correct way (according to manual test) :+1: 

I left a few inline comments - mostly very small details.

The casual full test run passed on my FF and CI Chrome: :+1: 
However, we have two failing security tests now due to `id` and `role` attribute was added to `li` elements.

Could you also adjust those two?

**Comment by Comandeer** [comment]
Adjusted tests, all should be green now.

**Review by sculpt0r — APPROVED** [review]
Good job :tada: LGTM :tada: 

I have to remove a single quote from the unit test message, which was... my suggestion before :see_no_evil: 

I verified corrected unit tests and they pass :+1:
